sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java

```diff
@@ -0,0 +1,465 @@
+        poolId,
+        read.getStreamName(),
+        consumerArn,
+        checkpoints);
+    for (ShardCheckpoint shardCheckpoint : checkpoints) {
+      checkState(
+          !state.containsKey(shardCheckpoint.getShardId()),
+          "Duplicate shard id %s",
+          shardCheckpoint.getShardId());
+      ShardState shardState = new ShardState(initShardSubscriber(shardCheckpoint), shardCheckpoint);
+      state.put(shardCheckpoint.getShardId(), shardState);
+    }
+  }
+
+  /**
+   * Returns the next disaggregated {@link KinesisRecord} if available and updates {@link #state}
+   * accordingly so that it reflects a mutable checkpoint AFTER returning that record.
+   *
+   * <p>Async subscription errors are delayed until {@link #eventQueue} is completely drained and
+   * then rethrown here.
+   *
+   * <p>This repeats the following steps until a record or {@code null} was returned:
+   *
+   * <ol>
+   *   <li>If {@link #current} is null and {@link #eventQueue} is empty, return {@code null} unless
+   *       {@link #subscriptionError} is set: in that case rethrow.
+   *   <li>Otherwise if {@link #current} is null, poll next from {@link #eventQueue}.
+   *   <li>If {@link #current} has a next {@link KinesisClientRecord}, update {@link #state}
+   *       accordingly and return the corresponding converted {@link KinesisRecord}, optionally
+   *       triggering {@link #onEventDone} if that was the last record of {@link #current}.
+   *   <li>Finally, if nothing was returned yet, trigger {@link #onEventDone} and continue loop.
+   * </ol>
+   *
+   * <p>It polls the {@link #eventQueue} in a while loop to avoid returning null immediately if an
+   * event without records arrived. There may be events with records after the {@link #current}, and
+   * it is better to poll again instead of having {@link EFOKinesisReader#advance()} signalling
+   * false to Beam. Otherwise, Beam would poll again later, which would introduce unnecessary delay.
+   */
+  @Nullable
+  KinesisRecord getNextRecord() throws IOException {
+    while (true) {
+      if (!isStopped && subscriptionError != null) {
+        // Stop the pool to cancel all subscribers and prevent new subscriptions.
+        // Doing this as part of getNextRecord() avoids concurrent access to the state map and
+        // prevents any related issues.
+        stop();
+      }
+
+      if (current == null) {
+        current = eventQueue.poll();
+      }
+
+      if (current != null) {
+        String shardId = current.shardId;
+        ShardState shardState = Preconditions.checkStateNotNull(state.get(shardId));
+        if (current.hasNext()) {
+          KinesisClientRecord r = current.next();
+          KinesisRecord kinesisRecord = new KinesisRecord(r, read.getStreamName(), shardId);
+          if (shardState.recordWasNotCheckPointedYet(kinesisRecord)) {
+            shardState.update(r);
+            watermarkPolicy.update(kinesisRecord);
```

Comment:
The way it's currently done here means that calling `getWatermark()` on the reader might yield non-monotonic values, yes. I did not know it might be an issue.

> Or, alternatively we track watermarks per shard and return the min watermark of these. As far as I know that's how it's done in the non EFO source.

Sure, will do that.